src/pubsub/ua_pubsub_ns0.c

```diff
@@ -1633,6 +1634,101 @@ removeDataSetWriterAction(UA_Server *server,
 }
 #endif
 
+#if defined(UA_ENABLE_PUBSUB_SKS) && defined(UA_ENABLE_PUBSUB_INFORMATIONMODEL_METHODS)
+static UA_StatusCode
+setSecurityKeysAction(UA_Server *server, const UA_NodeId *sessionId, void *sessionHandle,
+                      const UA_NodeId *methodId, void *methodContext,
+                      const UA_NodeId *objectId, void *objectContext, size_t inputSize,
+                      const UA_Variant *input, size_t outputSize, UA_Variant *output) {
+
+    /*Check whether the channel is encrypted according to specification*/
+    session_list_entry *session_entry;
+    LIST_FOREACH(session_entry, &server->sessions, pointers) {
+        if(UA_NodeId_equal(&session_entry->session.sessionId, sessionId)) {
+            if(session_entry->session.header.channel->securityMode !=
+               UA_MESSAGESECURITYMODE_SIGNANDENCRYPT)
+                return UA_STATUSCODE_BADSECURITYMODEINSUFFICIENT;
+        }
+    }
+
+    if(!server || !input)
+        return UA_STATUSCODE_BADINVALIDARGUMENT;
+
+    /* TODO: Check user credentials*/
```

Comment:
This comment can leave people confused whether this is inherently insecure as an implementation.
Rather write that the user credentials are checked in the AccessControl plugin *before* this callback is executed.
It is up to the user to configure his AccessControl plugin in a way that suits the use case.

Reply:
I added a @note before this callback and removed the comment.